File: src/main/ruby/g0101_0200/s0105_construct_binary_tree_from_preorder_and_inorder_traversal/solution.rb
# #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Hash_Table #Tree #Binary_Tree
# #Divide_and_Conquer #Data_Structure_II_Day_15_Tree #Big_O_Time_O(N)_Space_O(N)
# #2023_11_23_Time_71_ms_(91.78%)_Space_212_MB_(79.45%)

# Definition for a binary tree node.
# class TreeNode
#     attr_accessor :val, :left, :right
#     def initialize(val = 0, left = nil, right = nil)
#         @val = val
#         @left = left
#         @right = right
#     end
# end
# @param {Integer[]} preorder
# @param {Integer[]} inorder
# @return {TreeNode}
def build_tree(preorder, inorder)
  @j = 0
  @map = {}
  inorder.each_with_index {|val, index| @map[val] = index}
  answer(preorder, inorder, 0, preorder.length - 1)
end

private

def get(key)
  @map[key]
end

def answer(preorder, inorder, start, end_)
  return nil if start > end_ || @j > preorder.length

  value = preorder[@j]
  index = get(value)
  @j += 1
  node = TreeNode.new(value)
  node.left = answer(preorder, inorder, start, index - 1)
  node.right = answer(preorder, inorder, index + 1, end_)
  node
end

File: src/main/js/g0101_0200/s0105_construct_binary_tree_from_preorder_and_inorder_traversal/solution.js
// #Medium #Top_100_Liked_Questions #Top_Interview_Questions #Array #Hash_Table #Tree #Binary_Tree
// #Divide_and_Conquer #Data_Structure_II_Day_15_Tree #Top_Interview_150_Binary_Tree_General
// #Big_O_Time_O(N)_Space_O(N) #2024_12_13_Time_1_ms_(98.60%)_Space_53.9_MB_(86.87%)

import { TreeNode } from '../../com_github_leetcode/treenode';

/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} preorder
 * @param {number[]} inorder
 * @return {TreeNode}
 */
var buildTree = function(preorder, inorder) {
    let j = 0
    const map = new Map()

    for (let i = 0; i < inorder.length; i++) {
        map.set(inorder[i], i)
    }

    const build = (start, end) => {
        if (start > end || j >= preorder.length) {
            return null
        }

        const value = preorder[j++]
        const index = map.get(value)
        const node = new TreeNode(value)

        node.left = build(start, index - 1)
        node.right = build(index + 1, end)

        return node
    }

    return build(0, inorder.length - 1)
};

export { buildTree }
